compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp

```diff
@@ -0,0 +1,98 @@
+#include "mpk_fault_handler.h"
+#include "alloc_site_handler.h"
+#include "mpk.h"
+
+#include <sys/mman.h>
+
+// Set to whatever the default page size will be for page based MPK.
+#define PAGE_SIZE 4096
+
+// Trap Flag
+#define TF 0x100
+
+uint32_t last_pkey = INVALID_PKEY;
+unsigned int last_access_rights = PKEY_DISABLE_ACCESS;
+
+void disableMPK(siginfo_t *si, void *arg);
+
+void segMPKHandle(int sig, siginfo_t *si, void *arg) {
+  if (si->si_code != SEGV_PKUERR) {
+    __sanitizer::Report("INFO : SegFault other than SEGV_PKUERR, handling with "
+                        "default handler.\n");
+    // SignalHandler was invoked from an error other than MPK violation.
+    // Perform default action instead and return.
+    signal(sig, SIG_DFL);
+    raise(sig);
+    return;
+  }
+
+  // Obtains pointer causing fault
+  void *ptr = si->si_addr;
+
+  // Obtains the faulting pkey (in SEGV_PKUERR faults)
+  uint32_t pkey = si->si_pkey;
+
+  // Get Alloc Site information from the handler.
+  auto handler = __mpk_untrusted::AllocSiteHandler::init();
+  handler->addFaultAlloc((rust_ptr)ptr, pkey);
+  __sanitizer::Report(
+      "INFO : Got Allocation Site (%d) for address: %p with pkey: %d or %d.\n",
+      handler->getAllocSite((rust_ptr)ptr).id(), ptr, pkey);
+
+  // Logic for segfault handling separated out for
+  // easier switching between implementation strategies.
+  disableMPK(si, arg);
+}
+
+void disablePageMPK(siginfo_t *si, void *arg) {
+  // TODO : Not sure if this is the correct way to get the page, need to double check.
```

Comment:
The page address will allways be the pointer address without the last n-bits. So you can just mask them off. the number of bits is tied to the page size obviously generally `mask = PAGE_SIZE -1` so `page_start = ptr &~mask` where you're just masking off the bottom pointer bits.  will require `uintptr_t` or equivalent.